### PR TITLE
Gnutls support

### PR DESCRIPTION
**Pull request by babelouest**
Hi @benmcollins ,

Here is a work-in-progress for GnuTLS support of libjwt.

All test pass except for the ECDSA signatures, because I couldn't find how to convert to raw format.

The other signatures (SHA and RSA) are working fine according to the tests.

I didn't change the `configure.ac` file, I'm saying that you should know how to do it. The minimum version is GnuTLS 3.0.

### CODE REVIEW

File: libjwt/jwt-gnutls.c [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+      pk_alg = GNUTLS_PK_EC;
+      format = GNUTLS_X509_FMT_PEM;
+      break;
+    default:
+      alg = GNUTLS_DIG_NULL;
+      pk_alg = GNUTLS_PK_UNKNOWN;
+      format = GNUTLS_X509_FMT_PEM;
+      break;
+  }
+
```
Comment (benmcollins) [inline libjwt/jwt-gnutls.c]: Can you please break out this deeply nested if to single if's. This is too hard to follow and doesn't really serve a purpose to have it nested like this.

File: libjwt/jwt-gnutls.c [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  gnutls_datum_t key_dat = {(void *) jwt->key, jwt->key_len}, body_dat = {(void*) str, strlen(str)}, sig_dat;
+  int res, pk_alg;
+  int alg;
+  gnutls_x509_crt_fmt_t format;
+  
+  switch (jwt->alg) {
+    case JWT_ALG_RS256:
+      alg = GNUTLS_DIG_SHA256;
+      pk_alg = GNUTLS_PK_RSA;
+      format = GNUTLS_X509_FMT_PEM;
```
Comment (benmcollins) [inline libjwt/jwt-gnutls.c]: Looks like every one of these is `GNUTLS_X509_FMT_PEM`, so maybe just set that in the variable declaration block?.

File: libjwt/jwt-gnutls.c [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+      alg = GNUTLS_DIG_SHA384;
+      pk_alg = GNUTLS_PK_EC;
+      format = GNUTLS_X509_FMT_PEM;
+      break;
+    case JWT_ALG_ES512:
+      alg = GNUTLS_DIG_SHA512;
+      pk_alg = GNUTLS_PK_EC;
+      format = GNUTLS_X509_FMT_PEM;
+      break;
+    default:
```
Comment (benmcollins) [inline libjwt/jwt-gnutls.c]: The default should be `return EINVAL`. No reason to set random data just to check it later. It's a failure.

File: libjwt/jwt-gnutls.c [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    case JWT_ALG_HS256:
+      alg = GNUTLS_DIG_SHA256;
+      break;
+    case JWT_ALG_HS384:
+      alg = GNUTLS_DIG_SHA384;
+      break;
+    case JWT_ALG_HS512:
+      alg = GNUTLS_DIG_SHA512;
+      break;
+    default:
```
Comment (benmcollins) [inline libjwt/jwt-gnutls.c]: The default should be `return EINVAL`. No reason to set random data just to check it later. It's a failure.

File: libjwt/jwt-gnutls.c [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+  }
+  
+  if (alg != GNUTLS_DIG_NULL) {
+    *len = gnutls_hmac_get_len(alg);
+    (*out) = malloc(*len);
+    if (*out != NULL) {
+      if (!gnutls_hmac_fast(alg, jwt->key, jwt->key_len, str, strlen(str), (*out))) {
+        return 0;
+      } else {
+        return EINVAL;
```
Comment (benmcollins) [inline libjwt/jwt-gnutls.c]: Again, remove the nested if statements. If malloc fails (out == NULL) it should return ENOMEM.

File: libjwt/jwt-gnutls.c (hunk 4 above)
Comment (babelouest) [inline libjwt/jwt-gnutls.c]: done

File: libjwt/jwt-gnutls.c (hunk 5 above)
Comment (babelouest) [inline libjwt/jwt-gnutls.c]: done

File: libjwt/jwt-gnutls.c (hunk 2 above)
Comment (babelouest) [inline libjwt/jwt-gnutls.c]: yup, my bad, I was testing something for ECDSA signatures that didn't work, fixed

File: libjwt/jwt-gnutls.c (hunk 3 above)
Comment (babelouest) [inline libjwt/jwt-gnutls.c]: done

File: libjwt/jwt-gnutls.c (hunk 1 above)
Comment (babelouest) [inline libjwt/jwt-gnutls.c]: done

File: libjwt/jwt-gnutls.c [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+	/* Sign data */
+	if (gnutls_privkey_sign_data(privkey, alg, 0, &body_dat, &sig_dat)) {
+		res = EINVAL;
+		goto CLEAN_PRIVKEY;
+	}
+	
+	(*out) = malloc(sig_dat.size);
+	if (*out == NULL) {
+		res = ENOMEM;
+	}
```
Comment (benmcollins) [inline libjwt/jwt-gnutls.c]: I think you're missing a goto here.

File: libjwt/jwt-gnutls.c [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+int jwt_verify_sha_pem(jwt_t *jwt, const char *head, const char *sig_b64) {
+	int sig_len;
+	char * sig = jwt_b64_decode(sig_b64, &sig_len);
+	gnutls_datum_t sig_dat = {(void*)sig, sig_len}, cert_dat = {(void*)jwt->key, jwt->key_len}, data = {(void*)head, strlen(head)};
+	gnutls_pubkey_t pubkey;
+	int alg;
+	int res;
+	
+	switch (jwt->alg) {
+  case JWT_ALG_RS256:
```
Comment (benmcollins) [inline libjwt/jwt-gnutls.c]: All of these switch/case statements have bad indentation still.

File: libjwt/jwt-gnutls.c [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    break;
+  case JWT_ALG_ES384:
+    alg = GNUTLS_DIG_SHA384;
+    break;
+  case JWT_ALG_ES512:
+    alg = GNUTLS_DIG_SHA512;
+    break;
+  default:
+    return EINVAL;
+    break;
```
Comment (benmcollins) [inline libjwt/jwt-gnutls.c]: No need for break after return (same for above switch/case statements as well).

**Comment by babelouest** [comment]
The code is updated, no nested `if`, `switch/case` simplified and cleaned, and no more space instead of tabs

File: libjwt/jwt-gnutls.c [hunk 9]
```diff
@@ -0,0 +1,223 @@
+/* Copyright (C) 2017 Nicolas Mora <mail@babelouest.org>
+	 This file is part of the JWT C Library
```
Comment (benmcollins) [inline libjwt/jwt-gnutls.c]: I think you were too quick with the find/replace here. The copyright doesn't need tab indents. Just copy what's in `jwt.c` please.

File: libjwt/jwt-gnutls.c [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+	/* Sign data */
+	if (gnutls_privkey_sign_data(privkey, alg, 0, &body_dat, &sig_dat)) {
+		res = EINVAL;
+		goto CLEAN_PRIVKEY;
+	}
+	
+	(*out) = malloc(sig_dat.size);
+	if (*out == NULL) {
+		res = ENOMEM;
+	}
```
Comment (babelouest) [inline libjwt/jwt-gnutls.c]: sorry for the deletion, goto fixed

File: libjwt/jwt-gnutls.c (hunk 9 above)
Comment (babelouest) [inline libjwt/jwt-gnutls.c]: my bad, fixed

File: libjwt/jwt-gnutls.c (hunk 6 above)
Comment (babelouest) [inline libjwt/jwt-gnutls.c]: fixed

File: libjwt/jwt-gnutls.c (hunk 7 above)
Comment (babelouest) [inline libjwt/jwt-gnutls.c]: should be fixed now

File: libjwt/jwt-gnutls.c (hunk 8 above)
Comment (babelouest) [inline libjwt/jwt-gnutls.c]: fixed

**Comment by babelouest** [comment]
Hello @benmcollins ,

I've made improvements in the ECDSA signatures and I think it works better now. Also I've fixed the ECDSA tests that didn't always have the good length. All tests pass for gnutls and openssl. Let me know if you see something else.

**Comment by babelouest** [comment]
Another thing, GnuTLS 3.5 minimum is required, because those required functions don't exist in previous versions:

```c
int _gnutls_encode_ber_rs_raw(gnutls_datum_t * sig_value, const gnutls_datum_t * r, const gnutls_datum_t * s);
int _gnutls_decode_ber_rs_raw(const gnutls_datum_t * sig_value, gnutls_datum_t * r, gnutls_datum_t * s); 
```

**Comment by benmcollins** [comment]
Those look like internal functions, which I assume are not declared in any gnutls headers. My concern is they are not guaranteed to maintain any API/ABI compatibility, and could break in weird ways.

**Comment by babelouest** [comment]
You're right, I understand your concern, nevertheless, these functions are planned to be exported as public in GnuTLS 3.6:
https://gitlab.com/gnutls/gnutls/issues/205

Nikos suggestion is to replicate them until 3.6. I prefer using the internal functions instead.

Therefore, the options I see are:
- Use them like this, then adapt the code with the public functions when GnuTLS 3.6 will be released
- Don't use them and release libjwt with GnuTLS but without ECDSA signatures, they are not required by the RFC, but would be a very good feature. I don't know how often ECDSA signatures are used. In my case, HMAC and RSA are good enough, but I'm not everyone
- Don't accept the pull request and wait for GnuTLS with official support for encode/decode ECDSA signatures

I leave the final decision to you.

**Comment by benmcollins** [comment]
The problem is, the newly exported functions in 3.6 are named differently, so your current implementation will only work with 3.5. I suggest for < 3.6 (I assume GnuTLS has some define for the version at compile time), you add the simple wrappers that will then be compatible with 3.6+, e.g.:

    extern int _gnutls_encode_ber_rs_raw(gnutls_datum_t * sig_value, const gnutls_datum_t * r,
                                                                    const gnutls_datum_t * s);
    static int
    gnutls_encode_rs_value(gnutls_datum_t * sig_value,
                                            const gnutls_datum_t * r,
                                            const gnutls_datum_t * s)
    {
        return _gnutls_encode_ber_rs_raw(sig_value, r, s);
    }

Then use the new functions in your code. This will make it forward compatible and alleviate any API/ABI issues.

**Comment by babelouest** [comment]
You're correct, here is my new proposition.
Specific for the branch 3.5, assuming these functions will be public in GnuTLS 3.6.

I also identified and fixed a random bug due to padding missing when exported `r` or `s` were not long enough.

**Comment by benmcollins** [comment]
So I've pulled your jwt-gnutls.c file and reworked it...a lot. There were some possible memory leaks, lots of formatting issues and some awkward logic.

Everything looks good now. I also fixed the assumption that the curve bit depth needed to match the SHA. The number in ES256/384/512 equates to the SHA and not the curve bit (e.g. you can use a p384 EC key for ES256 and ES512).

I've also implemented all of the hooks in the build environment. By default it will look for OpenSSL, unless you pass `--without-openssl` in which case it will look for GnuTLS.

All tests pass either enabled.

**Comment by babelouest** [comment]
Good, thanks!
